src/main/java/com/fasterxml/jackson/core/util/ObjectPool.java

```diff
@@ -0,0 +1,330 @@
+    }
+
+    class StrategyHolder {
+        private static Strategy strategy = Strategy.LINKED_QUEUE;
+
+        public static void setStrategy(String name) {
+            strategy = Strategy.valueOf(name);
+        }
+    }
+
+    static <T> ObjectPool<T> newLockFreePool(Supplier<T> factory) {
+        switch (StrategyHolder.strategy) {
+            case DUMMY: return new DummyPool<>(factory);
+            case BUFFER_RECYCLERS: return new DummyPool<>(() -> (T) BufferRecyclers.getBufferRecycler());
+            case LINKED_QUEUE: return new LinkedQueuePool<>(factory);
+            case ATOMIC_REF_ARRAY_16: return new AtomicReferenceArrayPool<>(factory, 16);
+            case ATOMIC_REF_ARRAY_64: return new AtomicReferenceArrayPool<>(factory, 64);
+            case SYNCED_ARRAY_16: return new SyncedArrayPool<>(factory, 16);
+            case SYNCED_ARRAY_64: return new SyncedArrayPool<>(factory, 64);
+            case SHARDED_SYNCED_ARRAY_4_64: return new ShardedSyncedArrayPool<>(factory, 4, 64);
+            case SHARDED_SYNCED_ARRAY_16_16: return new ShardedSyncedArrayPool<>(factory, 16, 16);
+            case LOCK_FREE: return new LockFreePool<>(factory);
+            case LOCK_FREE_ASYNC_OFFER: return new LockFreeAsyncOfferPool<>(factory);
+            case JCTOOLS_16: return new JCToolsPool<>(factory, 16);
+            case JCTOOLS_64: return new JCToolsPool<>(factory, 64);
+        }
+        throw new UnsupportedOperationException();
+    }
+
+    class LinkedQueuePool<T> implements ObjectPool<T> {
+        private final Supplier<T> factory;
+        private final Consumer<T> destroyer;
+
+        private final Queue<T> pool = new LinkedTransferQueue<>();
+
+        public LinkedQueuePool(Supplier<T> factory) {
+            this(factory, null);
+        }
+
+        public LinkedQueuePool(Supplier<T> factory, Consumer<T> destroyer) {
+            this.factory = factory;
+            this.destroyer = destroyer;
+        }
+
+        @Override
+        public T borrow() {
+            T t = pool.poll();
+            return t != null ? t : factory.get();
+        }
+
+        @Override
+        public void offer(T t) {
+            pool.offer(t);
+        }
+
+        @Override
+        public void close() throws Exception {
+            if (destroyer != null) {
+                pool.forEach(destroyer);
+            }
+        }
+    }
+
+    class AtomicReferenceArrayPool<T> implements ObjectPool<T> {
+        private final Supplier<T> factory;
+
+        private final AtomicInteger counter = new AtomicInteger(0);
+
+        private final int size;
+
+        private final AtomicReferenceArray<T> pool;
+
+        public AtomicReferenceArrayPool(Supplier<T> factory, int size) {
+            this.factory = factory;
+            this.size = size;
+            this.pool = new AtomicReferenceArray<>(size);
+        }
+
+        @Override
+        public T borrow() {
+            int pos = counter.getAndUpdate(i -> i > 0 ? i-1 : i);
```

Comment:
Same performance problem of the Lock free variant pool: `counter` is highly contended (they are both stacks, actually - linked or array-based) hence would limit scalability. 
I would instead use a striped variant of this

eg

add a parameter "concurrency" which represent the number of "slots" of your pool.
each slot contains a fixed size partition of the total size (ie total size/concurrency).
If you make both total size and concurrency power of 2 would be better (more info later why).

You can then use the thread id to decide which slot to hit first (using power of 2 size allow to perform modulus with `threadId & (size -1)`) and you have 2 strategies here:
- performing cas on the counter owning that slot's index, moving tot he next counter, if failing till trying with all `concurrency` counters (that means that we need to keep track "where" we have borrowed it, or we risk to unbalance the pool on release - searching for some available slot)
- just keep on trying (via a cas loop) on the same slot's index: thanks to the thread id distributions (with v threads in particular) you'll probably be lucky to spread the contention the same

To better deal with the contention, too, the counters per partition could be all stored into an AtomicLongArray at 8/16 long distance to each others, to keep each counter separated by 1 or 2 cache lines to avoid false-sharing.

Another interesting way to deal with this, could be by make the algorithm a bit unbalanced
ie in order to make progress we need to first to borrow right?
Then we could just perform `getAndIncrement` there....
- if we overflow (more then slot capacity) we just switch to a mix of get and compareAndSet (because no available stuff is in) or just allocate
- the release side instead can check what's the status of the counter and use get + compareAndSet: if we overflow it means we have exhausted the capacity, and we just need to decrease it by 1 from the max capacity